lib/delta-sync/file-sync.js

```diff
@@ -0,0 +1,1051 @@
+import { query, update, sparqlEscapeUri, sparqlEscapeString } from 'mu';
+import { querySudo, updateSudo } from '@lblod/mu-auth-sudo';
+import { escapeRDFTerm, storeError } from "../utils.js";
+import http from "http";
+import fs   from "fs";
+import path from "path";
+import {
+  DISABLE_FILE_INGEST,
+  PREFIXES,
+  TEMP_FILE_GRAPH,
+  SYNC_BASE_URL,
+  DOWNLOAD_FILE_PATH,
+  FILE_FOLDER,
+  INGEST_GRAPH,
+  TEMP_FILE_REMOVAL_GRAPH,
+  REMAPPING,
+  MU_CALL_SCOPE_ID_FILE_SYNC
+} from '../../config.js';
+
+/**
+ * Start the synchronisation of files. Scan the temporary graph, download files and push file metadata to the rest of the application.
+ * Runs different functions to accomplish subtasks.
+ *
+ * @public
+ * @async
+ * @function startFileSync
+ * @return {undefined} - No useful return value, side effects only
+ */
+export async function startFileSync() {
+  console.info(`DISABLE_FILE_INGEST: ${DISABLE_FILE_INGEST}`);
+  try {
+    if (!DISABLE_FILE_INGEST) {
+     await runFileSync(); 
+    }
+    else {
+      console.warn('Automated file ingest disabled');
+    }
+  }
+  catch (err) {
+    console.error("Synchronisation of files failed with error:", err);
+    //Store the error to the database, also include the stacktrace for better debugging (even though the stack trace might be useless when the code changes over time).
+    await storeError(err.message + "\n" + err.stack.toString());
+  }
+}
+
+/**
+ * Run the actual file synchronisation. It downloads files, puts them in the correct location, buffers file metadata until all is received, removes files if metadata is deleted, and also buffers this data.
+ * Calls different functions to accomplish subtasks.
+ *
+ * @private
+ * @async
+ * @function runFileSync
+ */
+async function runFileSync() {
+
+  ////Scan and process the temporary graph for downloading files
+
+  try {
+    //Download files when the vuuid is available
+    await downloadFiles();
+  }
+  catch (err) {
+    //Downloads can fail, so log the error, and continue
+    for (let e of err) {
+      console.error(e.message + "\n" + e.stack.toString());
+      await storeError(e.message + "\n" + e.stack.toString());
+    }
+  }
+
+  //Rename and move the files, remap the uri
+  try {
+    await remapFiles();
```

Comment:
This was my effort to keep the remapping as a separate process that can fail 'gracefully'. So if the remapping fails, the file is still safe on the consumer, but in the wrong place. Different approach?

Reply:
Not sure I see the benefit here. If you write the file immediately to the correct location, there is no need to introduce a second step which could potentially fail. I might miss something though

Reply:
Remapping also involves rewriting the URIs in the database. This can only be done once all the metadata has arrived, otherwise you would have to somehow track which triples are already remapped and which aren't. When the file is being downloaded, not all metadata has necessarily arrived. I wanted to isolate changing the URI (which contains the file path) and moving the file to that path in one step in order to avoid an inconsistent state.